cmd/gaia/app/app.go

```diff
@@ -238,6 +240,34 @@ func (app *GaiaApp) initChainer(ctx sdk.Context, req abci.RequestInitChain) abci
 		panic(err) // TODO find a way to do this w/o panics
 	}
 
+	if len(genesisState.Txs) > 0 {
+		for _, genTx := range genesisState.Txs {
+			var tx auth.StdTx
+			err = app.cdc.UnmarshalJSON(genTx, &tx)
+			if err != nil {
+				panic(err)
+			}
+			bz, err := app.cdc.MarshalBinary(tx)
+			if err != nil {
+				panic(err)
+			}
+			res := app.BaseApp.DeliverTx(bz)
+			if !res.IsOK() {
+				panic(res.Log)
+			}
+		}
+
+		validators = app.stakeKeeper.ApplyAndReturnValidatorSetUpdates(ctx)
+		app.slashingKeeper.AddValidators(ctx, validators)
+	}
+
+	// sanity check
+	if len(req.Validators) > 0 {
+		if len(req.Validators) != len(validators) {
```

Comment:
Reading this I think it should panic? `RequestInitChain` doesn't know about the validators, they're set according to the genesis transactions.

Reply:
I couldn't quite get that - it does panic, do you think that it shouldn't?

Reply:
I understand this now in the context of your below code which sets the validators in the JSON file, but I don't understand why we need to set the validators in the JSON file at all.

Reply:
You don't need to, this is only enforcing that if you do, they better be the same as the result of running the txs.  Hence the `if len(req.Validators) > 0 { `

I guess you could just get rid of this and just error if `len(req.Validators) > 0`